apiserver/facades/client/modelupgrader/controllerupgrader.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelupgrader
+
+import (
+	"context"
+
+	"github.com/juju/names/v6"
+
+	"github.com/juju/juju/apiserver/authentication"
+	"github.com/juju/juju/apiserver/common"
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/apiserver/facade"
+	"github.com/juju/juju/core/agentbinary"
+	coreerrors "github.com/juju/juju/core/errors"
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/core/semversion"
+	controllerupgradererrors "github.com/juju/juju/domain/controllerupgrader/errors"
+	"github.com/juju/juju/domain/modelagent"
+	modelagenterrors "github.com/juju/juju/domain/modelagent/errors"
+	"github.com/juju/juju/internal/errors"
+	"github.com/juju/juju/rpc/params"
+)
+
+// ControllerUpgraderService mirrors the controller upgrader service
+// so we can easily mock it for unit tests.
+type ControllerUpgraderService interface {
+	// UpgradeController will upgrade the current clusters set of controllers to the latest
+	// patch version within the current controller's major and minor version.
+	UpgradeController(ctx context.Context) (semversion.Number, error)
+	// UpgradeControllerWithStream upgrades the current clusters set of controllers
+	// to the latest Juju version available using the given stream.
+	// It will be updated latest to the latest patch version within the current
+	// controller's major and minor version.
+	UpgradeControllerWithStream(
+		ctx context.Context,
+		stream modelagent.AgentStream,
+	) (semversion.Number, error)
+	// UpgradeControllerToVersion upgrades the current clusters set of controllers
+	// to the specified version.
+	UpgradeControllerToVersion(
+		ctx context.Context,
+		desiredVersion semversion.Number,
+	) error
+	// UpgradeControllerToVersionAndStream upgrades the current clusters set of
+	// controllers to the specified version. It will grab the binaries that is based
+	// on the given stream.
+	UpgradeControllerToVersionAndStream(
+		ctx context.Context,
+		desiredVersion semversion.Number,
+		stream modelagent.AgentStream,
+	) error
+
+	// CanUpgradeController determines whether the controller can be upgraded
+	// to the latest available patch version.
+	CanUpgradeController(ctx context.Context) (semversion.Number, error)
```

Comment:
By positioning all of these func names as `CanDoThingX` it implies that they should be returning a boolean. I'll read the rest of this PR before first.